### PR TITLE
Blog: Build a Rails app with TiDB

### PR DESCRIPTION
**Pull request by septemberfd**
### What is changed, added or deleted? (Required)

A new blog post is ready to release on 30 September.

### CODE REVIEW

File: build-a-ruby-on-rails-app-with-tidb [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+If you already know how to build a Rails app from scratch, you can skip steps 2-4 and start playing with our [demo application](https://github.com/hooopo/rails-tidb). 
+
+### **Step 1: Set up your local TiDB server**
+
+Deploy a TiDB cluster on your local machine.
+
+1. Install TiUP.
+
+TiDB provides a smooth deployment experience using [TiUP](https://docs.pingcap.com/tidb/dev/tiup-overview?_ga=2.62761701.816161674.1631694409-365976877.1630559498), a package manager for you to manage different cluster components easily in the TiDB ecosystem.
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb:37]:
```suggestion
TiDB provides a smooth deployment experience using [TiUP](https://docs.pingcap.com/tidb/dev/tiup-overview), a package manager for you to manage different cluster components easily in the TiDB ecosystem. 
```
Please check other links in this file and make sure they are correct.

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+title: Power Up Your Rails Apps with a NewSQL Database
+author: ['Matt Wang']
+date: 2020-09-30
+summary: This post helps Ruby on Rails developers get started with TiDB and use it as the backend storage layer of Rails applications.
+tags: ['Tutorial']
+categories: ['Community']
+image: /images/blog/build-rails-apps-with-tidb-banner.png
+---
+
+**Author:** [Matt Wang](https://github.com/zhongzc) (Engineer at PingCAP, moderator of Ruby-China community)
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:11]:
```suggestion
**Author:** [Matt Wang](https://github.com/hooopo) (Engineer at PingCAP, moderator of Ruby-China community) 
```
zhongzc is another colleague. Why is he here? 😂

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+![Build a Rails App with TiDB](media/build-rails-apps-with-tidb-banner.png)
+
+If you are a Ruby on Rails developer, I think you’ll really enjoy this article. It aims to help you get started with [TiDB](https://github.com/pingcap/tidb), an open-source NewSQL database, and use it to power up your Rails applications. 
+
+## Use TiDB to build up your Ruby on Rails applications
+
+[TiDB](https://github.com/pingcap/tidb) is an open-source NewSQL database that supports Hybrid Transactional and Analytical Processing (HTAP) workloads. It is MySQL compatible and features horizontal scalability, strong consistency, and high availability.
+
+To build apps with Rails, TiDB offers you MySQL interfaces which can be used as the backend database layers. TiDB allows you to use Active Record Object Relational Mapping (ORM) directly, and also provides you with an alternative: the [activerecord-tidb-adapter](https://github.com/pingcap/activerecord-tidb-adapter), a lightweight extension of Active Record that offers compatible patches and TiDB-exclusive functions such as[Sequence](https://docs.pingcap.com/tidb/stable/sql-statement-create-sequence).
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:23]:
```suggestion
To build apps with Rails, TiDB offers you MySQL interfaces which can be used as the backend database layers. TiDB allows you to use Active Record Object Relational Mapping (ORM) directly, and also provides you with an alternative: the [activerecord-tidb-adapter](https://github.com/pingcap/activerecord-tidb-adapter), a lightweight extension of Active Record that offers compatible patches and TiDB-exclusive functions such as [Sequence](https://docs.pingcap.com/tidb/stable/sql-statement-create-sequence). 
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+### Step 1: Set up your local TiDB server
+
+Deploy a TiDB cluster on your local machine.
+
+1. Install TiUP.
+
+TiDB provides a smooth deployment experience using [TiUP](https://docs.pingcap.com/tidb/dev/tiup-overview), a package manager for you to manage different cluster components easily in the TiDB ecosystem. 
+
+```shell
+
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:40]:
```suggestion
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+If you already know how to build a Rails app from scratch, you can skip steps 2-4 and start playing with our [demo application](https://github.com/hooopo/rails-tidb). 
+
+### Step 1: Set up your local TiDB server
+
+Deploy a TiDB cluster on your local machine.
+
+1. Install TiUP.
+
+TiDB provides a smooth deployment experience using [TiUP](https://docs.pingcap.com/tidb/dev/tiup-overview), a package manager for you to manage different cluster components easily in the TiDB ecosystem.
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:37]:
```suggestion
TiDB provides a smooth deployment experience using [TiUP](https://docs.pingcap.com/tidb/stable/tiup-overview), a package manager for you to manage different cluster components easily in the TiDB ecosystem. 
```

Comment (ran-huang) [reply build-a-ruby-on-rails-app-with-tidb.md:37]: Link to `stable` docs if applicable.

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+$ curl --proto '=https' --tlsv1.2 -sSf https://tiup-mirrors.pingcap.com/install.sh | sh
+```
+
+2. Start TiDB playground.
+
+Start a TiDB nightly instance by running the `tiup playground` command:  
+
+```shell
+
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:49]:
```suggestion
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+1. Make sure that you have Ruby and Rails installed, and  initiate a Rails app named `tidb-rails`; also be sure to set the database as `mysql` because TiDB speaks the MySQL protocol.
+
+```
+$ ruby -v
+
+ruby 2.7.0
+
+$ rails -v
+
+Rails 6.1.4
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:70]:
```suggestion
$ ruby -v
ruby 2.7.0

$ rails -v
Rails 6.1.4
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+$ rails new tidb-rails --database=mysql --api
+```
+
+2. Add[activerecord-tidb-adapter](https://github.com/pingcap/activerecord-tidb-adapter) to Gemfile. Activerecord-tidb-adapter allows you to use TiDB as a backend for ActiveRecord and Rails apps.
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:75]:
```suggestion
2. Add [activerecord-tidb-adapter](https://github.com/pingcap/activerecord-tidb-adapter) to Gemfile. Activerecord-tidb-adapter allows you to use TiDB as a backend for ActiveRecord and Rails apps.
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+default: &default
+
+ adapter: tidb
+
+ encoding: utf8mb4
+
+ collation: utf8mb4_general_ci
+
+ pool: &lt;%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:93]:
```suggestion
 pool:  <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+   tidb_enable_noop_functions: ON
+
+ username: root
+
+ password:
+
+development:
+
+ &lt;<: *default
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:109]:
```suggestion
 <<: *default
```

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+ database: tidb_rails_development
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:111]: Please remove the extra blank lines.

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+
+Create the local database for your rails application.
+
+```
+$ bundle exec rails db:create
+
+Created database 'tidb_rails_development'
+
+Created database 'tidb_rails_test'
+```
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:126]: ditto

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+     t.integer :gender
+
+     t.timestamps
+
+   end
+
+ end
+
+end
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:162]: ditto

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+```
+$ bundle exec rails db:migrate
+
+== 20210826174523 CreateUsers: migrating ======================================
+
+-- create_table(:users)
+
+  -> 0.1717s
+
+== 20210826174523 CreateUsers: migrated (0.1717s) =============================
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:176]: ditto

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+
+  (8.9ms)  SELECT COUNT(*) FROM `users`
+
+=> 30
+
+irb(main):003:0> User.first
+
+ User Load (5.8ms)  SELECT `users`.* FROM `users` ORDER BY `users`.`id` ASC LIMIT 1
+
+=> #&lt;User id: 1, email: "user-0@example.com", name: "user-0", gender: 0, created_at: "2021-08-26 17:50:40.661945000 +0000", updated_at: "2021-08-26 17:50:40.661945000 +0000">
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:212]:
```suggestion
=> #<User id: 1, email: "user-0@example.com", name: "user-0", gender: 0, created_at: "2021-08-26 17:50:40.661945000 +0000", updated_at: "2021-08-26 17:50:40.661945000 +0000">
```

File: build-a-ruby-on-rails-app-with-tidb [hunk 16]
```diff
@@ -0,0 +1,219 @@
+---
+title: Build a Ruby on Rails App with TiDB and the Active Record TiDB Adapter
```
[the remark this answers is not in the document]
Comment (septemberfd) [reply build-a-ruby-on-rails-app-with-tidb]: ok

File: build-a-ruby-on-rails-app-with-tidb.md (hunk 11 above)
Comment (septemberfd) [reply build-a-ruby-on-rails-app-with-tidb.md]: ok

File: build-a-ruby-on-rails-app-with-tidb.md (hunk 12 above)
Comment (septemberfd) [reply build-a-ruby-on-rails-app-with-tidb.md:122]: ok

File: build-a-ruby-on-rails-app-with-tidb.md (hunk 13 above)
Comment (septemberfd) [reply build-a-ruby-on-rails-app-with-tidb.md:158]: ok

File: build-a-ruby-on-rails-app-with-tidb.md (hunk 14 above)
Comment (septemberfd) [reply build-a-ruby-on-rails-app-with-tidb.md:172]: ok

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+tags: ['Tutorial']
+categories: ['Community']
+image: /images/blog/build-rails-apps-with-tidb-banner.png
+---
+
+**Author:** [Matt Wang](https://github.com/hooopo) (Engineer at PingCAP, moderator of Ruby-China community)
+
+**Editor:** [Fendy Feng](https://github.com/septemberfd), Tom Dewan
+
+![Build a Rails App with TiDB](media/build-rails-apps-with-tidb-banner.png)
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:15]:
```suggestion
![Build a Rails App with a NewSQL Database](media/build-rails-apps-with-a-newsql-database.png)
```
We can insert as many keywords as possible in image names and alt text for SEO purposes. Please change the image file name accordingly.

**Review by ran-huang — COMMENTED** [review]
Rest LGTM

**Review by ran-huang — APPROVED** [review]

File: build-a-ruby-on-rails-app-with-tidb.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
+---
+title: Power Up Your Rails Apps with a NewSQL Database
+author: ['Matt Wang']
+date: 2020-09-30
+summary: This post helps Ruby on Rails developers get started with TiDB and use it as the backend storage layer of Rails applications.
+tags: ['Tutorial']
+categories: ['Community']
+image: /images/blog/build-rails-apps-with-tidb-banner.png
```
Comment (ran-huang) [inline build-a-ruby-on-rails-app-with-tidb.md:8]:
```suggestion
image: /images/blog/build-rails-apps-with-a-newsql-database.png
```

**Pull request merged**